scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,370 @@
+ * a certain time.
+ *
+ * The enqueued/dispatched queues are used in a producer/consumer way between
+ * the BPF part and the user-space part.
+ */
+#define MAX_ENQUEUED_TASKS 1024
+
+/*
+ * Tasks enqueued to the user-space for scheduling.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_QUEUE);
+	__uint(key, 0);
+	__type(value, struct task_ctx);
+	__uint(max_entries, MAX_ENQUEUED_TASKS);
+} enqueued SEC(".maps");
+
+/*
+ * Tasks enqueued by the user-space for dispatching.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_QUEUE);
+	__uint(key, 0);
+	__type(value, struct task_ctx);
+	__uint(max_entries, MAX_ENQUEUED_TASKS);
+} dispatched SEC(".maps");
+
+/* Per-task local storage */
+struct task_storage {
+	u64 start_ns; /* Updated when the task is running */
+	u64 stop_ns;  /* Updated when the task is stopping */
+};
+
+/* Map that contains task-local storage. */
+struct {
+	__uint(type, BPF_MAP_TYPE_TASK_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct task_storage);
+} task_storage_map SEC(".maps");
+
+/* Return true if the target task @p is a kernel thread */
+static inline bool is_kthread(const struct task_struct *p)
+{
+	return !!(p->flags & PF_KTHREAD);
+}
+
+/* Return true if the target task @p is the user-space scheduler */
+static inline bool is_usersched_task(const struct task_struct *p)
+{
+	return p->pid == usersched_pid;
+}
+
+/* Return a (wrapping safe) time delta */
+static inline u64 time_diff(u64 end, u64 start)
+{
+	return (s64)end - (s64)start;
+}
+
+/* Update task timestamps in the local task storage */
+static void update_task_time(struct task_struct *p,
+			     bool update_start, bool update_stop)
+{
+	struct task_storage *ts;
+	u64 now = bpf_ktime_get_ns();
+
+	ts = bpf_task_storage_get(&task_storage_map, p, 0, 0);
+	if (!ts)
+		return;
+	if (update_start)
+		ts->start_ns = now;
+	if (update_stop)
+		ts->stop_ns = now;
+}
+
+/* Task @p starts to run on a CPU */
+void BPF_STRUCT_OPS(rustlite_running, struct task_struct *p)
+{
+	bpf_printk("start: pid=%d (%s)", p->pid, p->comm);
+	update_task_time(p, true, false);
+}
+
+/* Task @p releases a CPU */
+void BPF_STRUCT_OPS(rustlite_stopping, struct task_struct *p, bool runnable)
+{
+	bpf_printk("stop: pid=%d (%s)", p->pid, p->comm);
+	update_task_time(p, false, true);
+}
+
+/*
+ * Return the time slice (in ns) used by the task during the previous execution
+ * slot (up to a maximum of slice_ns = full time slice used).
+ */
+static u64 task_slice_used(struct task_struct *p)
+{
+	struct task_storage *ts;
+	u64 time_ns;
+
+	ts = bpf_task_storage_get(&task_storage_map, p, 0, 0);
+	if (!ts)
+		return 0;
+	time_ns = time_diff(ts->stop_ns, ts->start_ns);
+
+	return MIN(time_ns, slice_ns);
+}
+
+/*
+ * Return task's vtime, evaluated as following:
+ *
+ *    current time + used task slice (scaled based on task's priority)
+ */
+static u64 task_vtime(struct task_struct *p)
+{
+	return bpf_ktime_get_ns() + task_slice_used(p) * 100 / p->scx.weight;
```

Comment:
Good point. I've used wallclock time here, because I thought it was a bit more intuitive. But I like the idea of reporting the total used cputime to userspace and move the whole vtime logic to userspace as well.

That's the whole point here, the idea is to provide something like a "scheduling lab" where people can easily experiment their policies simply by changing the Rust code (without touching any kernel/BPF code).

So +1 on that.